File: src/validators/xml-validator.ts
import { XmlParser } from '../core/xml-parser';
import { SettingsDocument, XmlFile, ValidationResult, Setting } from '../types';

export class XmlValidator {
  private xmlParser: XmlParser;

  constructor() {
    this.xmlParser = new XmlParser();
  }

  validateXmlFiles(xmlFiles: XmlFile[], settingsDocument: SettingsDocument): ValidationResult[] {
    const results: ValidationResult[] = [];

    for (const xmlFile of xmlFiles) {
      for (const setting of settingsDocument.settings) {
        const result = this.validateSetting(xmlFile, setting);
        results.push(result);
      }
    }

    return results;
  }

  private validateSetting(xmlFile: XmlFile, setting: Setting): ValidationResult {
    try {
      const actualValue = this.xmlParser.evaluateXPath(xmlFile.content, setting.xpath);
      
      if (actualValue === null || actualValue === undefined) {
        return {
          settingName: setting.name,
          xpath: setting.xpath,
          expectedValue: setting.expectedValue,
          actualValue: null,
          status: 'missing',
          message: `Setting not found at XPath: ${setting.xpath}`,
          filePath: xmlFile.path
        };
      }

      const convertedActualValue = this.convertValue(actualValue, setting.type || 'string');
      const isValid = this.compareValues(convertedActualValue, setting.expectedValue, setting.type);

      return {
        settingName: setting.name,
        xpath: setting.xpath,
        expectedValue: setting.expectedValue,
        actualValue: convertedActualValue,
        status: isValid ? 'pass' : 'fail',
        message: isValid 
          ? undefined 
          : `Expected ${setting.expectedValue}, got ${convertedActualValue}`,
        filePath: xmlFile.path
      };
    } catch (error) {
      return {
        settingName: setting.name,
        xpath: setting.xpath,
        expectedValue: setting.expectedValue,
        actualValue: null,
        status: 'fail',
        message: `Validation error: ${error instanceof Error ? error.message : String(error)}`,
        filePath: xmlFile.path
      };
    }
  }

  private convertValue(value: string, type: string): string | number | boolean {
    switch (type) {
      case 'number':
        const numValue = parseFloat(value);
        if (isNaN(numValue)) {
          throw new Error(`Cannot convert "${value}" to number`);
        }
        return numValue;
      case 'boolean':
        const lowerValue = value.toLowerCase().trim();
        if (lowerValue === 'true' || lowerValue === '1') return true;
        if (lowerValue === 'false' || lowerValue === '0') return false;
        throw new Error(`Cannot convert "${value}" to boolean`);
      case 'string':
      default:
        return value.trim();
    }
  }

  private compareValues(
    actualValue: string | number | boolean, 
    expectedValue: string | number | boolean | undefined,
    type?: string
  ): boolean {
    if (expectedValue === undefined) {
      return true;
    }

    if (type === 'number') {
      return Number(actualValue) === Number(expectedValue);
    }

    if (type === 'boolean') {
      return Boolean(actualValue) === Boolean(expectedValue);
    }

    return String(actualValue) === String(expectedValue);
  }
}

File: xml_inspector/validators/xml_validator.py
"""XML validation functionality."""

from typing import List, Union
import logging

from ..core.xml_parser import XmlParser, XPathEvaluationError
from ..types import (
    SettingsDocument, 
    XmlFile, 
    ValidationResult, 
    Setting, 
    SettingValue
)

logger = logging.getLogger(__name__)


class ValidationError(Exception):
    """Raised when validation fails."""
    pass


class XmlValidator:
    """Validates XML files against settings documents."""
    
    def __init__(self) -> None:
        """Initialize the XML validator."""
        self.xml_parser = XmlParser()
    
    def validate_xml_files(
        self, 
        xml_files: List[XmlFile], 
        settings_document: SettingsDocument
    ) -> List[ValidationResult]:
        """
        Validate multiple XML files against a settings document.
        
        Args:
            xml_files: List of XmlFile objects to validate
            settings_document: Settings document containing validation rules
            
        Returns:
            List of ValidationResult objects
        """
        results = []
        
        for xml_file in xml_files:
            for setting in settings_document.settings:
                result = self._validate_setting(xml_file, setting)
                results.append(result)
        
        return results
    
    def _validate_setting(self, xml_file: XmlFile, setting: Setting) -> ValidationResult:
        """
        Validate a single setting against an XML file.
        
        Args:
            xml_file: XML file to validate
            setting: Setting definition to validate against
            
        Returns:
            ValidationResult object
        """
        try:
            actual_value_str = self.xml_parser.evaluate_xpath(
                xml_file.content, 
                setting.xpath
            )
            
            if actual_value_str is None:
                return ValidationResult(
                    setting_name=setting.name,
                    xpath=setting.xpath,
                    expected_value=setting.expected_value,
                    actual_value=None,
                    status="missing",
                    message=f"Setting not found at XPath: {setting.xpath}",
                    file_path=xml_file.path
                )
            
            try:
                actual_value = self._convert_value(actual_value_str, setting.type)
            except ValueError as e:
                return ValidationResult(
                    setting_name=setting.name,
                    xpath=setting.xpath,
                    expected_value=setting.expected_value,
                    actual_value=actual_value_str,
                    status="fail",
                    message=f"Type conversion error: {e}",
                    file_path=xml_file.path
                )
            
            is_valid = self._compare_values(actual_value, setting.expected_value, setting.type)
            
            return ValidationResult(
                setting_name=setting.name,
                xpath=setting.xpath,
                expected_value=setting.expected_value,
                actual_value=actual_value,
                status="pass" if is_valid else "fail",
                message=None if is_valid else f"Expected {setting.expected_value}, got {actual_value}",
                file_path=xml_file.path
            )
            
        except XPathEvaluationError as e:
            return ValidationResult(
                setting_name=setting.name,
                xpath=setting.xpath,
                expected_value=setting.expected_value,
                actual_value=None,
                status="fail",
                message=f"XPath evaluation error: {e}",
                file_path=xml_file.path
            )
        except Exception as e:
            return ValidationResult(
                setting_name=setting.name,
                xpath=setting.xpath,
                expected_value=setting.expected_value,
                actual_value=None,
                status="fail",
                message=f"Validation error: {e}",
                file_path=xml_file.path
            )
    
    def _convert_value(self, value: str, value_type: str) -> SettingValue:
        """
        Convert string value to appropriate type.
        
        Args:
            value: String value to convert
            value_type: Target type ('string', 'number', 'boolean')
            
        Returns:
            Converted value
            
        Raises:
            ValueError: If conversion fails
        """
        value = value.strip()
        
        if value_type == "number":
            try:
                # Try integer first
                if '.' not in value:
                    return int(value)
                else:
                    return float(value)
            except ValueError:
                raise ValueError(f'Cannot convert "{value}" to number')
        
        elif value_type == "boolean":
            lower_value = value.lower()
            if lower_value in ['true', '1', 'yes', 'on']:
                return True
            elif lower_value in ['false', '0', 'no', 'off']:
                return False
            else:
                raise ValueError(f'Cannot convert "{value}" to boolean')
        
        else:  # string or default
            return value
    
    def _compare_values(
        self, 
        actual_value: SettingValue, 
        expected_value: Union[SettingValue, None], 
        value_type: str
    ) -> bool:
        """
        Compare actual and expected values.
        
        Args:
            actual_value: Actual value from XML
            expected_value: Expected value from settings
            value_type: Type of comparison to perform
            
        Returns:
            True if values match or no expected value is specified
        """
        if expected_value is None:
            return True
        
        if value_type == "number":
            # Handle number comparison with type conversion
            try:
                actual_num = float(actual_value) if actual_value is not None else 0
                expected_num = float(expected_value)
                return actual_num == expected_num
            except (ValueError, TypeError):
                return False
        
        elif value_type == "boolean":
            # Handle boolean comparison
            return bool(actual_value) == bool(expected_value)
        
        else:  # string comparison
            return str(actual_value) == str(expected_value)